[NuGet.Versioning] SemanticVersion.HasMetadata should indicate that Metadata is not null when true

### NuGet Product Used

NuGet SDK

### Product Version

NuGet.Versioning v6.7.0

### Worked before?

no

### Impact

It's more difficult to complete my work

### Repro Steps & Context

1. `dotnet new console -f net6.0 -o NugetVersioningHasMetadataRepro`
2. `cd NugetVersioningHasMetadataRepro`
3. `dotnet add package NuGet Nuget.Versioning --version 6.7.0`
4. In `Program.cs` write:
    ```csharp
    var x = NuGet.Versioning.SemanticVersion.Parse("1.0.0");
    var output = x.HasMetadata ? x.Metadata[0..Math.Min(4, x.Metadata.Length)] : string.Empty;
    ```

What I expected:
5. Build with no warnings.

What I got:
5. Observe build warnings on `x.Metadata`: `Program.cs(2,30): warning CS8602: Dereference of a possibly null reference.`




### Verbose Logs

_No response_

--- a/src/NuGet.Core/NuGet.Versioning/SemanticVersion.cs
+++ b/src/NuGet.Core/NuGet.Versioning/SemanticVersion.cs
@@ -4,6 +4,7 @@
 using System;
 using System.Collections.Generic;
 using System.ComponentModel;
+using System.Diagnostics.CodeAnalysis;
 using System.Linq;
 
 namespace NuGet.Versioning
@@ -231,6 +232,7 @@ namespace NuGet.Versioning
         /// <summary>
         /// True if metadata exists for the version.
         /// </summary>
+        [MemberNotNullWhen(true, nameof(Metadata))]
         public virtual bool HasMetadata
         {
             get { return !string.IsNullOrEmpty(Metadata); }

--- a/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionTests.cs
@@ -90,7 +90,7 @@ namespace NuGet.Versioning.Test
         [InlineData("1.2.3-A.00.B")]
         public void TryParseStrictReturnsFalseIfVersionIsNotStrictSemVer(string version)
         {
-            // Act 
+            // Act
             SemanticVersion? semanticVersion;
             var result = SemanticVersion.TryParse(version, out semanticVersion);
 
@@ -107,6 +107,20 @@ namespace NuGet.Versioning.Test
             string result = target.ToString();
 
             Assert.Equal("1.2.3", result);
+        }
+
+        [Fact]
+        public void Metadata_NoNullableWarning_After_HasMetadata_checked()
+        {
+            // Arrange
+            SemanticVersion target = new(1, 2, 3);
+
+            // Act
+            // should not result in a compiler warning CS8602: Dereference of a possibly null reference.
+            string result = target.HasMetadata ? target.Metadata[0..1] : "no-metadata";
+
+            // Assert
+            Assert.Equal("no-metadata", result);
         }
 
         private class ExtendedSemanticVersion : SemanticVersion

--- a/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Versioning.Test/SemanticVersionTests.cs
@@ -117,7 +117,7 @@ namespace NuGet.Versioning.Test
 
             // Act
             // should not result in a compiler warning CS8602: Dereference of a possibly null reference.
-            string result = target.HasMetadata ? target.Metadata[0..1] : "no-metadata";
+            string result = target.HasMetadata ? target.Metadata.Substring(1) : "no-metadata";
 
             // Assert
             Assert.Equal("no-metadata", result);

